File: old/cicada/src/env.rs
use std::sync::Arc;
use std::collections::HashMap;

use crate::val::Val;

#[derive(Clone)]
#[derive(Debug)]
#[derive(PartialEq, Eq)]
pub struct Env {
    val_map: HashMap<String, Arc<Val>>
}

impl Env {
    pub fn new() -> Env {
        let val_map = HashMap::new();
        Env { val_map }
    }

    pub fn lookup_val(&self, field: &String) -> Option<Arc<Val>> {
        match self.val_map.get(field) {
            Some(val) => {
                Some(val.clone())
            }
            None => {
                None
            }
        }
    }

    pub fn ext(&self, name: String, val: Arc<Val>) -> Env {
        let mut val_map = self.val_map.clone();
        val_map.insert(name, val);
        Env { val_map }
    }
}

File: old/src/main/scala/xieyuheng/cicada/Env.scala
package xieyuheng.cicada

case class Env(map: Map[String, Val] = Map()) {

  def lookup_val(name: String): Option[Val] = {
    map.get(name)
  }

  def ext(name: String, value: Val): Env = {
    Env(this.map + (name -> value))
  }

  def ext_map(map: Map[String, Val]): Env = {
    Env(this.map ++ map)
  }

}
